### ISSUE TITLE
reconstruction produces disjoint path

### ISSUE THREAD

**Issue by fhill122** [colmap/colmap#1521]
Hi there,

I'll just start with the picture to illustrate my issue:
![Pasted image 20220511172012](https://user-images.githubusercontent.com/13213165/167818918-872df8f7-9771-4c81-8f94-953bd8466b20.png)

All images are actually taken from a **single continuous trajectory**, but reconstruction always seems failed to connect the other images when registering the last one. 
Note this is **not a loop closure issue**, this happens in the middle of the trajectory. 

The matching is done with sequential matching, and seems the **gapped images have pretty good  matching**:
![Pasted image 20220511172120](https://user-images.githubusercontent.com/13213165/167820324-2a254661-4eac-4643-ba53-f32d52d2dd89.png)

Any idea where it goes wrong?
Appreciate the help!

**Comment by tsattler** [comment]
>  Note this is not a loop closure issue, this happens in the middle of the trajectory.

I think that it is a loop closure problem. Colmap can add images in arbitrary order. In this case, it looks like Colmap used some of the images on the right side (e.g., one of the images from the straight trajectory and the other one from the trajectory that was turning the corner) as Colmap prefers to use image pairs that are not pure forward motion for initialization. Likely, it continued adding some images from the turning trajectory and then continued with images from the straight trajectory. I guess the latter trajectory will at some point turn left and come back, creating a loop closure. But by the time Colmap "returns" to the starting position, it has accumulated too much drift to connect both ends of the trajectory, leading to the disjoint path you observe.

**Comment by fhill122** [comment]
> I think that it is a loop closure problem. Colmap can add images in arbitrary order. In this case, it looks like Colmap used some of the images on the right side (e.g., one of the images from the straight trajectory and the other one from the trajectory that was turning the corner) as Colmap prefers to use image pairs that are not pure forward motion for initialization. Likely, it continued adding some images from the turning trajectory and then continued with images from the straight trajectory. I guess the latter trajectory will at some point turn left and come back, creating a loop closure. But by the time Colmap "returns" to the starting position, it has accumulated too much drift to connect both ends of the trajectory, leading to the disjoint path you observe.

thanks tsattler. I have added the overall image, the whole trajectory is as what you suspected. 

Apart from tuning to improve the accuracy and reducing the drift, do you have any suggestion to help connect the trajectory? would adding gps helps?
I am confused especially for the part that this happened  even though **the image pair have good two-view geometries matching**. To my understanding, this could be done through a sim3 pose graph optimization followed by a global optimization. Is this a missing feature for colmap?

**Comment by Serenitysmk** [comment]
> > I think that it is a loop closure problem. Colmap can add images in arbitrary order. In this case, it looks like Colmap used some of the images on the right side (e.g., one of the images from the straight trajectory and the other one from the trajectory that was turning the corner) as Colmap prefers to use image pairs that are not pure forward motion for initialization. Likely, it continued adding some images from the turning trajectory and then continued with images from the straight trajectory. I guess the latter trajectory will at some point turn left and come back, creating a loop closure. But by the time Colmap "returns" to the starting position, it has accumulated too much drift to connect both ends of the trajectory, leading to the disjoint path you observe.
> 
> thanks tsattler. I have added the overall image, the whole trajectory is as what you suspected.
> 
> Apart from tuning to improve the accuracy and reducing the drift, do you have any suggestion to help connect the trajectory? would adding gps helps? I am confused especially for the part that this happened even though **the image pair have good two-view geometries matching**. To my understanding, this could be done through a sim3 pose graph optimization followed by a global optimization. Is this a missing feature for colmap?

Yes, explicitly finding a loop and closing the loop (redistributing the loop error in the entire trajectory ) is a missing feature at the moment. But actually, it is fairly straightforward to add GPS constraint in bundle adjustment which can help you to reduce drift in the beginning, but still, there is no loop closure module yet.

**Comment by eduardohenriquearnold** [comment]
Maybe too late, but in case anyone else runs into this:

@fhill122 assuming you do have "loop-closure" matches in the database, i.e. matches between images in the beginning/end of the sequence, you can give [GLOMAP](https://github.com/colmap/glomap) a try. It leverages joint optimization of all images rotations/translation, which avoids the drift introduced by incremental mapping.

In my experience, it was able to fix the drift from COLMAP:
![Screenshot from 2024-10-31 16-55-15](https://github.com/user-attachments/assets/14dac59d-2786-4858-bbf9-207309e9e2cd)

into the trajectory I'd expected:
![image](https://github.com/user-attachments/assets/9e985a30-928d-47ec-81f8-2911e6dcaab0)
